baseapp/baseapp.go

```diff
@@ -609,6 +609,141 @@ func (app *BaseApp) cacheTxContext(ctx sdk.Context, txBytes []byte) (sdk.Context
 	return ctx.WithMultiStore(msCache), msCache
 }
 
+// runCheckTxConcurrently processes a transaction with either the checkTx or recheckTx modes, encoded transaction
+// bytes, and the decoded transaction itself. All state transitions occur through
+// a cached Context depending on the mode provided.
+//
+// Note, gas execution info is always returned. A reference to a Result is
+// returned if the tx does not run out of gas and if all the messages are valid
+// and execute successfully. An error is returned otherwise.
+func (app *BaseApp) runCheckTxConcurrently(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, result *sdk.Result, anteEvents []abci.Event, priority int64, err error) {
+	if mode != runTxModeCheck && mode != runTxModeReCheck {
+		panic("runCheckTxConcurrently can only be invoked for CheckTx and RecheckTx.")
+	}
+
+	// Strip out the post handler
+	if app.postHandler != nil {
+		panic("CheckTx/RecheckTx does not support a post hander.")
+	}
+
+	// NOTE: GasWanted should be returned by the AnteHandler. GasUsed is
+	// determined by the GasMeter. We need access to the context to get the gas
+	// meter, so we initialize upfront.
+	var gasWanted uint64
```

Comment:
IIUC, gas is only considered when we are in `deliverTx` mode. Is this the correct understanding?

If so, do we need to do any gas related work here (i.e. setting `gasWanted` and recovery func below)?

Reply:
> IIUC, gas is only considered when we are in `deliverTx` mode. Is this the correct understanding?
 
We check that the transaction has enough gas during `CheckTx` and consume the gas during `DeliverTx`. We fail the `CheckTx` if it wouldn't have enough gas.

> If so, do we need to do any gas related work here (i.e. setting `gasWanted` and recovery func below)?

Yes, we have to return any gas computations from CheckTx back to the caller.